master/static/srv/ship_logs.py

```diff
@@ -0,0 +1,598 @@
+# Excluding empty spaces this regex matches rank in the above example as [rank=0]
+rank = re.compile(r"(?P<space1> ?)\[rank=(?P<rank_id>([0-9]+))\](?P<space2> ?)(?P<log>.*)")
+# Below regex is used to extract the message severity from the log message.
+# Excluding empty spaces and delimiter(:) this regex matches message severity level in the above
+# example as INFO
+level = re.compile(
+    r"(?P<space1> ?)(?P<level>(DEBUG|INFO|WARNING|ERROR|CRITICAL)):(?P<space2> ?)(?P<log>.*)"
+)
+lineend = re.compile(rb"[\r\n]")
+
+
+# Interval at which to force a flush.
+SHIPPER_FLUSH_INTERVAL = 1  # How often to make API calls
+
+# Full jitter time on encountering an API exception.
+SHIPPER_FAILURE_BACKOFF_SECONDS = 1
+
+# Max size of the log buffer before forcing a flush.
+LOG_BATCH_MAX_SIZE = 1000
+
+# Max size of the shipping queue before we start to apply backpressure by blocking sends. We would
+# only hit this if we got underwater by three full batches while trying to ship a batch.
+SHIP_QUEUE_MAX_SIZE = 3 * LOG_BATCH_MAX_SIZE
+
+
+def read_newlines_or_carriage_returns(fd: io.RawIOBase) -> Iterator[str]:
+    r"""
+    Read lines, delineated by either '\n' or '\r.
+
+    Unlike the default io.BufferedReader used in subprocess.Popen(bufsize=-1), we read until we
+    encounter either '\n' or \r', and treat that as one line.
+
+    Specifically, io.BufferedReader doesn't handle tqdm progress bar outputs very well; it treats
+    all of the '\r' outputs as one enormous line.
+
+    Args:
+        fd: an unbuffered stdout or stderr from a subprocess.Popen.
+
+    Yields:
+        A series of str, one per line.  Each line always ends with a '\n'.  Each line will be
+        broken to length io.DEFAULT_BUFFER_SIZE, even if the underlying io didn't have a linebreak.
+    """
+    # Ship lines of length of DEFAULT_BUFFER_SIZE, including the terminating newline.
+    limit = io.DEFAULT_BUFFER_SIZE - 1
+    nread = 0
+    chunks = []  # type: List[bytes]
+
+    def oneline():
+        nonlocal nread
+        nonlocal chunks
+        out = b"".join(chunks).decode("utf8")
+        chunks = []
+        nread = 0
+        return out
+
+    while True:
+        buf = fd.read(limit - nread)
+        if not buf:
+            # EOF.
+            break
+
+        # Extract all the lines from this buffer.
+        while buf:
+            m = lineend.search(buf)
+            if m is None:
+                # No line break here; just append to chunks.
+                chunks.append(buf)
+                nread += len(buf)
+                break
+
+            # Line break found!
+            start, end = m.span()
+            chunks.append(buf[:start])
+            # Even if we matched a '\r', emit a '\n'.
+            chunks.append(b"\n")
+            yield oneline()
+            # keep checking the rest of buf
+            buf = buf[end:]
+
+        # Detect if we reached our buffer limit.
+        if nread >= limit:
```

Comment:
A good portion of my own time checking this function was spent doing math with `nread`. But I think `nread` might not even be needed. Could this be simply:
```python
if len(chunks) == limit:
```
and could the read higher up be
```python
buf = fd.read(limit - len(chunks))
```
?

Then while we're at it, maybe `chunks` doesn't even have to exist. I think maybe the read could instead be:
```python
buf.append(fd.read(limit - len(buf))
```

Reply:
During my run yesterday after work, I found myself wondering about the arguments passed to `fd.read`. Maybe it's better to just `fd.read(limit)` every time? Or `fd.read(limit/2)`, or... I don't know. It's just not obvious to me that something like `fd.read(limit - nread)` is necessary or even better.

Reply:
This is "read up to the amount we're willing to buffer".

If you want to document a buffer size, reading more than that doesn't make sense, it just complicates the code.  It does result in potentially more system calls under certain circumstances, but I think that effect is unlikely to ever matter unless people are emitting a lot of enormous lines.

But I do think it's worth it to stick to a fixed buffer size because it's much easier to document and understand for users than "well if you lines are longer than X we reserve the right to break them at any point up to length Y" which is the behavior we'd have otherwise.

Since `chunks` is a list of bytestrings, `len(chunks)` is useless information; you'd have to replace `nread` with `sum(len(c) for c in chunks)` which feels pointless when you coudl instead just track an integer.